google/cloud/connection_options.h

```diff
@@ -0,0 +1,243 @@
+        user_agent_prefix_(ConnectionTraits::user_agent_prefix()),
+        background_threads_factory_(internal::DefaultBackgroundThreads) {
+    internal::DefaultLogging();
+  }
+
+  /// Change the gRPC credentials from the default of
+  /// `grpc::GoogleDefaultCredentials()`.
+  ConnectionOptions& set_credentials(
+      std::shared_ptr<grpc::ChannelCredentials> v) {
+    credentials_ = std::move(v);
+    return *this;
+  }
+
+  /// The gRPC credentials used by clients configured with this object.
+  std::shared_ptr<grpc::ChannelCredentials> credentials() const {
+    return credentials_;
+  }
+
+  /**
+   * Change the gRPC endpoint used to contact the Cloud Spanner service.
+   *
+   * In almost all cases the default ("spanner.googleapis.com") is the correct
+   * endpoint to use. It may need to be changed to (1) test against a fake or
+   * simulator, or (2) to use a beta or EAP version of the service.
+   */
+  ConnectionOptions& set_endpoint(std::string v) {
+    endpoint_ = std::move(v);
+    return *this;
+  }
+
+  /// The endpoint used by clients configured with this object.
+  std::string const& endpoint() const { return endpoint_; }
+
+  /**
+   * The number of transport channels to create.
+   *
+   * Some transports limit the number of simultaneous calls in progress on a
+   * channel (for gRPC the limit is 100). Increasing the number of channels
+   * thus increases the number of operations that can be in progress in
+   * parallel.
+   *
+   * The default value is 4.
+   */
+  int num_channels() const { return num_channels_; }
+
+  /// Set the value for `num_channels()`.
+  ConnectionOptions& set_num_channels(int num_channels) {
+    num_channels_ = num_channels;
+    return *this;
+  }
+
+  /**
+   * Return whether tracing is enabled for the given @p component.
+   *
+   * The Cloud Spanner C++ client can log interesting events to help library and
+   * application developers troubleshoot problems with the library or their
+   * configuration. This flag returns true if tracing should be enabled by
+   * clients configured with this option.
+   *
+   * Currently only the `rpc` component is supported, which enables logging of
+   * each RPC, including its parameters and any responses.
+   */
+  bool tracing_enabled(std::string const& component) const {
+    return tracing_components_.find(component) != tracing_components_.end();
+  }
+
+  /// Enable tracing for @p component in clients configured with this object.
+  ConnectionOptions& enable_tracing(std::string const& component) {
+    tracing_components_.insert(component);
+    return *this;
+  }
+
+  /// Disable tracing for @p component in clients configured with this object.
+  ConnectionOptions& disable_tracing(std::string const& component) {
+    tracing_components_.erase(component);
+    return *this;
+  }
+
+  /// Return the options for use when tracing RPCs.
+  TracingOptions tracing_options() const { return tracing_options_; }
```

Comment:
This may be my fault (or feature), but if all the other getters are going to return by `const&`, why not `tracing_options()` too (or vice versa).

Reply:
Fixed.